### PR TITLE
fix(VSelect): ensure VMenu's content has scroll listener added

### PR DESCRIPTION
**Pull request by bart5**
## Description
Ensure that `VSelect` adds scroll listener to `VMenu`'s `content`.

## Motivation and Context
Arbitrary one-time delay of `this.getContent().addEventListener('scroll', this.onScroll, false)` does not guarantee that target element will exist the next time that we try to add the listener.
Proposed solution does guarantee that listener is eventually added.

## How Has This Been Tested?
In the playground.

## Reproduction

```vue
<template>
  <v-container>
    <v-alert>Loaded {{ this.options.length }} options</v-alert>

    <v-combobox ref="combobox" :items="options" @click="firstLoad && loadOptions()">
      <template #append-item>
        <div v-intersect="onNextPageIntersection"/>
        <v-progress-linear :active="loading" indeterminate class="fixed-bottom"></v-progress-linear>
      </template>
    </v-combobox>
    <div class>1. See that no options are loaded. Open the menu.
      <br>2. You notice that some options loaded. Scroll down.
      <br>3. You can see that new batch of options loads yet we cannot see past the first 20.
      <br>
      <br>If you inspect `v-menu__content` you will notice that it has no `scroll` listener.
    </div>
  </v-container>
</template>

<script>
  export default {
    data: () => ({
      loading: false,
      options: [],
      firstLoad: true,
    }),
    methods: {
      onNextPageIntersection(entries, observer, isIntersecting) {
        if (isIntersecting && !this.loading) {
          this.loadOptions()
        }
      },
      loadOptions() {
        if (this.firstLoad) this.firstLoad = false
        this.loading = true
        setTimeout(() => {
          this.options.push(
            ...Array(20)
              .fill(0)
              .map(() => Math.round(Math.random() * 100000))
          )
          this.loading = false
        }, 1000)
      },
    },
  }
</script>

<style>
  .fixed-bottom {
    position: sticky !important;
    bottom: 8px;
    transform: translateY(8px);
  }

  .v-select-list {
    position: relative !important;
  }
</style>
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.

### CODE REVIEW

**Comment by KaelWD** [comment]
Duplicate of #11675

**Comment by KaelWD** [comment]
Aaand github won't let me reopen the original because it was force-pushed.

**Comment by bart5** [comment]
"I'm still not entirely sold on repeatedly listening to the updated event like that though, it feels kinda messy."
I agree. It is a trade-off. We pay several invocations of addListener for the certainty of it being effective eventually. In my tests I saw the update running 3 or 4 times before it could actually attach the listener. 

However, it seems to be the quickest and easiest solution, and frankly, pretty reliable.

I pushed another proposition with more direct approach.

**Comment by KaelWD** [comment]
What about just having VMenu emit the scroll event itself?

```diff
diff --git a/packages/vuetify/src/components/VMenu/VMenu.ts b/packages/vuetify/src/components/VMenu/VMenu.ts
index 83765e2ca..958d29aea 100644
--- a/packages/vuetify/src/components/VMenu/VMenu.ts
+++ b/packages/vuetify/src/components/VMenu/VMenu.ts
@@ -350,6 +350,7 @@ export default baseMixins.extend({
             if (this.closeOnContentClick) this.isActive = false
           },
           keydown: this.onKeyDown,
+          scroll: this.$listeners.scroll,
         },
       } as VNodeData
 
diff --git a/packages/vuetify/src/components/VSelect/VSelect.ts b/packages/vuetify/src/components/VSelect/VSelect.ts
index f120f0902..b55a90ae0 100644
--- a/packages/vuetify/src/components/VSelect/VSelect.ts
+++ b/packages/vuetify/src/components/VSelect/VSelect.ts
@@ -112,7 +112,6 @@ export default baseMixins.extend<options>().extend({
   data () {
     return {
       cachedItems: this.cacheItems ? this.items : [],
-      menuIsBooted: false,
       isMenuActive: false,
       lastItem: 20,
       // As long as a value is defined, show it
@@ -249,15 +248,6 @@ export default baseMixins.extend<options>().extend({
     },
     isMenuActive (val) {
       window.setTimeout(() => this.onMenuActiveChange(val))
-
-      if (!val) return
-
-      this.$nextTick(() => {
-        if (!this.menuIsBooted && this.getContent() && this.$refs.menu.isBooted) {
-          this.getContent().addEventListener('scroll', this.onScroll, false)
-          this.menuIsBooted = true
-        }
-      })
     },
     items: {
       immediate: true,
@@ -518,6 +508,7 @@ export default baseMixins.extend<options>().extend({
             this.isMenuActive = val
             this.isFocused = val
           },
+          scroll: this.onScroll,
         },
         ref: 'menu',
       }, [this.genList()])
```

**Comment by bart5** [comment]
I think it is by far the most sane solution.

File: packages/vuetify/src/components/VMenu/VMenu.ts [hunk 1]
```diff
@@ -350,6 +350,9 @@ export default baseMixins.extend({
             if (this.closeOnContentClick) this.isActive = false
           },
           keydown: this.onKeyDown,
+          ...this.$listeners.scroll
+            ? { scroll: this.$listeners.scroll }
```
Comment (KaelWD) [inline packages/vuetify/src/components/VMenu/VMenu.ts:354]: Does `undefined` do something bad?

Comment (bart5) [reply packages/vuetify/src/components/VMenu/VMenu.ts:354]: Vue throws because of it and there are some `TypeErrors` later along the way.

Comment (bart5) [reply packages/vuetify/src/components/VMenu/VMenu.ts:354]: We could `scroll: this.$listeners.scroll || false` (yes, false works :man_shrugging:) or `scroll: this.$listeners.scroll || (() => {})` but it leaves us with listener that does nothing.

Comment (bart5) [reply packages/vuetify/src/components/VMenu/VMenu.ts:354]: Alternatively:
```js
      if (this.$listeners.scroll) {
        options.on = options.on || {}
        options.on.scroll = this.$listeners.scroll
      }
```

**Comment by bart5** [comment]
We can also get rid of `scroll` listener completely:
```diff
diff --git a/packages/vuetify/src/components/VMenu/VMenu.ts b/packages/vuetify/src/components/VMenu/VMenu.ts
index a85202633..97b166f56 100644
--- a/packages/vuetify/src/components/VMenu/VMenu.ts
+++ b/packages/vuetify/src/components/VMenu/VMenu.ts
@@ -174,6 +174,10 @@ export default baseMixins.extend({
     },
     isContentActive (val) {
       this.hasJustFocused = val
+
+      if (!val) {
+        requestAnimationFrame(() => (this.$refs.content.scrollTop = 0))
+      }
     },
     listIndex (next, prev) {
       if (next in this.tiles) {
@@ -350,9 +354,6 @@ export default baseMixins.extend({
             if (this.closeOnContentClick) this.isActive = false
           },
           keydown: this.onKeyDown,
-          ...this.$listeners.scroll
-            ? { scroll: this.$listeners.scroll }
-            : {},
         },
       } as VNodeData
 
diff --git a/packages/vuetify/src/components/VSelect/VSelect.ts b/packages/vuetify/src/components/VSelect/VSelect.ts
index 5f84be822..818979ddd 100644
--- a/packages/vuetify/src/components/VSelect/VSelect.ts
+++ b/packages/vuetify/src/components/VSelect/VSelect.ts
@@ -217,7 +217,7 @@ export default baseMixins.extend<options>().extend({
     virtualizedItems (): object[] {
       return (this.$_menuProps as any).auto
         ? this.computedItems
         ? this.computedItems
-        : this.computedItems.slice(0, this.lastItem)
+        : this.computedItems.slice(0, this.getLastItem())
     },
     menuCanShow: () => true,
     $_menuProps (): object {
@@ -509,7 +509,6 @@ export default baseMixins.extend<options>().extend({
             this.isMenuActive = val
             this.isFocused = val
           },
-          scroll: this.onScroll,
         },
         ref: 'menu',
       }, [this.genList()])
@@ -625,7 +624,7 @@ export default baseMixins.extend<options>().extend({
       })
       const item = this.allItems[index]
       if (index !== -1) {
-        this.lastItem = Math.max(this.lastItem, index + 5)
+        this.lastItem = Math.max(this.getLastItem(), index + 5)
         this.setValue(this.returnObject ? item : this.getValue(item))
         this.$nextTick(() => this.$refs.menu.getTiles())
         setTimeout(() => this.setMenuIndex(index))
@@ -715,21 +714,19 @@ export default baseMixins.extend<options>().extend({
 
       VTextField.options.methods.onMouseUp.call(this, e)
     },
-    onScroll () {
-      if (!this.isMenuActive) {
-        requestAnimationFrame(() => (this.getContent().scrollTop = 0))
+    getLastItem (): number {
+      if (!this.isMenuActive || !this.getContent() || this.lastItem > this.computedItems.length) {
+        return this.lastItem
       } else {
-        if (this.lastItem > this.computedItems.length) return
-
         const showMoreItems = (
           this.getContent().scrollHeight -
           (this.getContent().scrollTop +
           this.getContent().clientHeight)
         ) < 200
-
         if (showMoreItems) {
           this.lastItem += 20
         }
+        return this.lastItem
       }
     },
     onSpaceDown (e: KeyboardEvent) {
```

**Comment by KaelWD** [comment]
How would it load more items when you scroll then?

**Comment by bart5** [comment]
As you can see above we can put `updateLastItem` into `items` watcher.
```diff
@@ -253,15 +253,15 @@ export default baseMixins.extend<options>().extend({
     items: {
       immediate: true,
       handler (val) {
...
```

**Comment by bart5** [comment]
Because Vuetify itself is not loading items it's enough to put `updateLastItem` into `items` watcher handler.
If items change `virtualizedItems`  will be changed accordingly.

Also, now thinking about it, we should add to that `updateLastItem` cap on `lastItem` in case the number of item decreased (e.g. new, narrower search). Otherwise if I gradually load e.g. 500 items scrolling down, then close the menu, run search, get some 400 results, we will try to render all those at once.

**Comment by bart5** [comment]
Actually, I think we should completely reset `lastItem` to default once menu is closed. Why keep all those nodes in the DOM?

**Comment by bart5** [comment]
I now realised that it doesn't work for lists loaded as a single batch.
I think it should be solved by delegating decision to the user whether their items list is paginated or not.

**Comment by KaelWD** [comment]
There's a feature request for proper pagination (#4867), anything we do here is still just a hacky workaround. 

> Why keep all those nodes in the DOM?

I think we've considered making v-menu remove itself from the DOM when closed.

> on reopening the menu, `lastItem` should be set back to `computedItems` length

You'd have to store the last value of `lastItem`, otherwise I could open the menu, see 20 items, close and reopen it and now I've got all 1000. 

We should just focus on the relevant bug here, `lastItem` needs to be replaced with the new virtual scroller anyway (#2660)

File: packages/vuetify/src/components/VMenu/VMenu.ts (hunk 1 above)
Comment (KaelWD) [reply packages/vuetify/src/components/VMenu/VMenu.ts:354]: Yeah I like that better, and we're already doing it for the mouse handlers.

File: packages/vuetify/src/components/VMenu/VMenu.ts [hunk 2]
```diff
@@ -349,10 +349,14 @@ export default baseMixins.extend({
             if (target.getAttribute('disabled')) return
             if (this.closeOnContentClick) this.isActive = false
           },
-          keydown: this.onKeyDown,
```
Comment (KaelWD) [inline packages/vuetify/src/components/VMenu/VMenu.ts:352]: ?

**Pull request merged**